### ISSUE TITLE
Versioned dev/build performance with large datasets

### ISSUE THREAD

**Issue by caffeineinc** [silverstripe/silverstripe-framework#8958]
## Affected Version

3.7.x

## Description

During a `dev/build` Silverstripe calls an extension `augmentDatabase()` on Versioned DataObjects.  During this function a cleanup query is made which was supposed to cleanup instances where the [uniqueness constraint was failed. ](https://github.com/silverstripe/silverstripe-framework/blob/3.7/model/Versioned.php#L542) Unfortunately this query is extremely slow for large data sets, and from the comments, it appears the original bug was fixed.  

The issue seems to have manifested in few different ways, which were resolved differently. Although i think the function may have been previously required, i'm not sure it now needs to run on every `dev/build`.

Instead of running this every time, it might more sense to provide the cleanup as a task, so that it can be run if necessary, while large datasets don't have the performance impact of running this _each_ `dev/build`

## Steps to Reproduce

- Create a versioned DataObject with 5-8 million rows
- Run a `/dev/build`

Related:
https://github.com/silverstripe/silverstripe-framework/issues/7944
https://github.com/silverstripe/silverstripe-versioned/issues/194

**Comment by caffeineinc** [comment]
This looks like it was resolved in 4.x.x after versioned `1.2.0`, since it was abstracted to a module - and you can flag disable the feature.

That said, i still don't think flag disabling it is a _great_ idea, since it still runs every build. I guess the bigger question is if it's still required. Larger websites might not run `dev/build` for a while, and having slow queries in there still don't make sense as many might not know about some hidden feature flag unless it is better documented. 

To me it feels like something that really should be be in a task like [this] (https://github.com/caffeineinc/silverstripe-framework/commit/8b14e21abd2a4c9acb2f3622132d28255394f5aa) pr, then it can be run at times when a slow query isn't so blocking..

**Comment by maxime-rainville** [comment]
I think this part of Versionned is now handling this uniqueness problem in SS4: https://github.com/silverstripe/silverstripe-versioned/blob/2d19b8bca39d73de221ca8cfc78100100172d828/src/Versioned.php#L892-L929

**Comment by maxime-rainville** [comment]
I'm not sure this would be high impact enough to warrant fixing in SS3. Unfortunately, [SilverStripe 3 has entered limited support in June 2018](https://www.silverstripe.org/blog/update-on-silverstripe-5-x/). This means we'll only be fixing critical bugs and security issues for SilverStripe 3 going forward.

We should take some time to make sure this logic is performant in SS4.

**Comment by caffeineinc** [comment]
Yeah, it's more the point [here](https://github.com/silverstripe/silverstripe-framework/issues/7944#issuecomment-376753996): 

> It'll just be a flag to allow this orphan check to be disabled.

The resolution was to add a _undocumented_ flag but the task is still executed each `dev/build` which means the problem still affects the majority of deployments, and every `dev/build` with a large `versioned` set, multiplied by every subclass.

I would say, by default the change should be *opt in* since that was [handled as you say](https://github.com/silverstripe/silverstripe-versioned/blob/2d19b8bca39d73de221ca8cfc78100100172d828/src/Versioned.php#L892-L929) Then we can document it, and if users still encounter issues, they can enable it if required. 

My question is that high impact enough to warrant a change like above for `silverstripe/versioned` ?

**Comment by caffeineinc** [comment]
Ok. so i looked closely at the new code in silverstripe versioned - It needs to be present in the case that deletes aren't cascaded, so i don't think we need to explicitly disable it as it might not have a performance impact like the fix for SS3. 

Could be worth merge but not a release ?

**Comment by GuySartorelli** [comment]
My understanding from the last comment is that there's no change to make for CMS 4+. Closing.
